yarn-project/ethereum/src/l1_tx_utils.ts

```diff
@@ -0,0 +1,309 @@
+    ...numberConfigHelper(3),
+  },
+  checkIntervalMs: {
+    description: 'How often to check tx status',
+    env: 'L1_TX_MONITOR_CHECK_INTERVAL_MS',
+    ...numberConfigHelper(30_000),
+  },
+  stallTimeMs: {
+    description: 'How long before considering tx stalled',
+    env: 'L1_TX_MONITOR_STALL_TIME_MS',
+    ...numberConfigHelper(60_000),
+  },
+};
+
+export const defaultL1TxUtilsConfig = getDefaultConfig<L1TxUtilsConfig>(l1TxUtilsConfigMappings);
+
+export interface L1TxRequest {
+  to: Address | null;
+  data: Hex;
+  value?: bigint;
+}
+
+interface GasPrice {
+  maxFeePerGas: bigint;
+  maxPriorityFeePerGas: bigint;
+}
+
+export class L1TxUtils {
+  private readonly config: L1TxUtilsConfig;
+
+  constructor(
+    private readonly publicClient: PublicClient,
+    private readonly walletClient: WalletClient<HttpTransport, Chain, Account>,
+    private readonly logger?: DebugLogger,
+    config?: Partial<L1TxUtilsConfig>,
+  ) {
+    this.config = {
+      ...defaultL1TxUtilsConfig,
+      ...(config || {}),
+    };
+  }
+
+  /**
+   * Sends a transaction with gas estimation and pricing
+   * @param request - The transaction request (to, data, value)
+   * @param gasConfig - Optional gas configuration
+   * @returns The transaction hash and parameters used
+   */
+  public async sendTransaction(
+    request: L1TxRequest,
+    _gasConfig?: Partial<L1TxUtilsConfig>,
+  ): Promise<{ txHash: Hex; gasLimit: bigint; gasPrice: GasPrice }> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const account = this.walletClient.account;
+
+    const gasLimit = await this.estimateGas(account, request);
+    const gasPrice = await this.getGasPrice(gasConfig);
+
+    const txHash = await this.walletClient.sendTransaction({
+      ...request,
+      gas: gasLimit,
+      maxFeePerGas: gasPrice.maxFeePerGas,
+      maxPriorityFeePerGas: gasPrice.maxPriorityFeePerGas,
+    });
+
+    this.logger?.verbose(
+      `Sent L1 transaction ${txHash} with gas limit ${gasLimit} and price ${formatGwei(gasPrice.maxFeePerGas)} gwei`,
+    );
+
+    return { txHash, gasLimit, gasPrice };
+  }
+
+  /**
+   * Monitors a transaction until completion, handling speed-ups if needed
+   * @param request - Original transaction request (needed for speed-ups)
+   * @param initialTxHash - Hash of the initial transaction
+   * @param params - Parameters used in the initial transaction
+   * @param gasConfig - Optional gas configuration
+   */
+  public async monitorTransaction(
+    request: L1TxRequest,
+    initialTxHash: Hex,
+    params: { gasLimit: bigint },
+    _gasConfig?: Partial<L1TxUtilsConfig>,
+  ): Promise<TransactionReceipt> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const account = this.walletClient.account;
+
+    const tx = await this.publicClient.getTransaction({ hash: initialTxHash });
+    if (tx?.nonce === undefined || tx?.nonce === null) {
+      throw new Error(`Failed to get L1 transaction ${initialTxHash} nonce`);
```

Comment:
This could happen if we try to monitor immediately after sending and the tx has not yet propagated. We should try to extract the nonce from the tx we've sent, instead of requesting it from the node. If we can't, let's add a small grace period before throwing.